examples/step-67/step-67.cc

```diff
@@ -2248,6 +2257,105 @@ namespace Euler_DG
+    //
+    // A few things need to be discussed before improving this dummy
+    // implementation:
+    // - Make PETSc Vector classes inherit from VectorSpaceVector
+    // - Decide on predefined PETScWrappers::TimeStepper constructors that also
+    //   select subtype
+    //   For example in this case we are only allowed to select Runge Kutta from
+    //   the code, but we need to use the command line to switch among the
+    //   various rk implementations in PETSc. The default RK solver in PETSc is
+    //   third order Bogacki-Shampine, with adaptive time step.
+    // - The deal.II monitor adapts the time step with the CFL estimate. PETSc
+    //   has a very similar API that requires downcasting (TSSetCFLTimeLocal).
+    //   How do you prefer to support this?
+    //   Note that the default adaptors for Runge Kutta with
+    //   embedded formulas produce steps very similar to the one with the CFL.
+    using VectorType = PETScWrappers::MPI::Vector;
+
+    // We won't need those ugly copy from/to operations once we derive PETSc
+    // vector classes from VectorSpaceVectors
+    auto copy_from_petsc = [](const VectorType &                          src,
+                              LinearAlgebra::distributed::Vector<Number> &dst) {
+      auto lr = src.local_range();
+      for (unsigned int i = lr.first; i < lr.second; i++)
+        dst[i] = src[i];
+    };
+
+    auto copy_to_petsc =
+      [](const LinearAlgebra::distributed::Vector<Number> &src,
+         VectorType &                                      dst) {
+        auto lr = dst.local_range();
+        for (unsigned int i = lr.first; i < lr.second; i++)
+          dst[i] = src[i];
+        dst.compress(VectorOperation::insert);
+      };
+
+    // Solver parameters. Here we specify initial time step and final time,
+    // together with the solver type for Runge-Kutta.
+    // We also set adaptive tolerances to -1 to use adaptive time stepping
+    // with default tolerances from PETSc
+    PETScWrappers::TimeStepperData tsdata;
+    tsdata.final_time         = final_time;
+    tsdata.initial_step_size  = time_step;
+    tsdata.tstype             = "rk";
+    tsdata.absolute_tolerance = -1.0;
+    tsdata.relative_tolerance = -1.0;
+
+    // Solver constructor
+    PETScWrappers::TimeStepper<VectorType> petsc_time_stepper(tsdata);
+
+    // PETScWrappers::TimeStepper can solve ODE/DAE problems in explicit form
+    // i.e., u_dot = G(u,t), or in fully implicit form F(u_dot,u,t) = 0.
+    // This function implements the explicit interface for G.
+    petsc_time_stepper.explicit_function =
+      [&](const double t, const VectorType &y, VectorType &f) {
+        // we need to set "time" of the Euler instance for boundary conditions
+        // evaluation
+        time = t;
+        copy_from_petsc(y, rk_register_1);
+        euler_operator.apply(t, rk_register_1, rk_register_2);
+        copy_to_petsc(rk_register_2, f);
+        return 0;
+      };
```

Comment:
This is an interesting direction, and I think it would be really helpful to some of our users to have this as an alternative (I am not sure how precisely we would offer such a functionality in terms of education and usefulness, as said elsewhere by @peterrum and @luca-heltai, so I have not something fundamentally now to contribute, apart from the fact that I would really like to have this alternative available). What would be of interest to me is the run times of the program with the two variants, let's say normalized to the number of operator evaluations. The performance is one of the primary areas of this tutorial programs, and I would be curious in how PETSc is doing along those lines, because this gives up a little performance.

Reply:
@kronbichler You can compare meaningful runtimes when these needless copies will be gone.